### PR TITLE
Various configure fixes and improvements

### LINKED ISSUE #912: Overly strict checking for MUMPS?
If the user asks for MUMPS support, we check for the following files during configure:

- dmumps_c.h 
- libdmumps.a 
- libmumps_common.a 
- libpord.a 
- libscalapack.a 
- libblacs.a 
- libparmetis.a 
- libmetis.a 
- libptscotch.a 
- libptesmumps.a

but I'm not sure we need all of these. The version of MUMPS in Ubuntu 14.04 (used on Travis), doesn't come with any of libscalapack.a, libblacs.a, libparmetis.a, or libmetis.a, so we turn off MUMPS support.

Do we know which of these are actually required?

### LINKED ISSUE #913: Configure checks not strict enough for SUNDIALS
We need `nvector_parallel.h`, which is not in the version of SUNDIALS available on Ubuntu 14 (used on Travis), but the configure checks pass.

This is because we check for `sundials-config`, query it for what flags to use and use them -- we don't check it has parallel capabilities. I'm not sure if this is possible, as `sundials-config` as removed in v2.6.0 when they moved to CMake. I've not established if some other config checker mechanism made it back in -- there doesn't appear to be anything in v2.7.0 though.

One option would be to try to compile a test program that uses everything we need, and turn off support if that doesn't work.

On a side-note, the whole configure section for MUMPS/CVODE/IDA/ARKODE/SUNDIALS needs a good tidy.

### PR DESCRIPTION
**Pull request by ZedThree**
I know, autoconf is terrible. This PR aims to generally tidy up and "improve" configure.

1. Replace "echo" with "AC_MSG_NOTICE" everywhere. This has two benefits:
    a) Adds the message to `config.log`
    b) Turns off the message if you run `configure` with `--quiet`
2. Fail immediately if an option couldn't be enabled for whatever reason and the user *explicitly* requested it, i.e. `./configure --with-foo --with-bar=/some/path` will fail as soon as possible if either "foo" or "bar" can't be found/enabled. Previously, some options would let configure carry on (somewhat) silently.
3. Sensible defaults. Lapack defaults to "guess" -- if Lapack can be found, great!, enable it. If it can't be found, don't worry about it.
4. Make things work on Archer. HDF5 couldn't be enabled on Archer because we relied on the `h5cc` wrapper script. Instead, we now just try to link a simple HDF5 program, which should work on Archer. Also applies to Lapack (maybe other options).
5. Allow explicitly turning off optimisations
6. Simplify some of the configuration. Try to make some bits easier to read (ha!), use existing autoconf macros rather than do things ourselves, cut out some cases we don't need to consider. For example, the SUNDIALS IDA/CVODE/ARKODE config stuff was all basically identical. So that's been moved into a macro, which is called three times with different arguments for the bits that need to be different. These are a bit less clever, and don't look in lots of different locations. The user will now need to explicitly pass the path if they aren't installed in a standard location. Fixes #913 
7. Move the configuration summary to be the very last thing, so it's hopefully on screen. Also removes the "octave" bit, as this is unmaintained.
8. Fixes #912 by allowing the user to provide `LIBS/LDFLAGS/CPPFLAGS` that should allow compiling/linking against MUMPS

### CODE REVIEW

**Comment by dschwoerer** [comment]
Cleaning `configure.ac` is hard work - thanks for that :+1: 
Just some minor comments:
* I prefer `."$enable_backtrace" == .maybe` over `x"$enable_backtrace" == xmaybe` - I think that makes it easier to read.
* Are the parallel versions of netcdf and hdf5 actually supported or could we remove them?

**Comment by d7919** [comment]
Is there any way to make the configure errors more obvious? The Travis test fails in the "maximal" coverage case due to asking for mumps but configure deciding it's not available. It takes some careful reading to spot the error lines in https://travis-ci.org/boutproject/BOUT-dev/jobs/353952927 (see line 1298).

**Comment by ZedThree** [comment]
Currently failing due to broken MUMPS

File: make.config.in [hunk 1]
```diff
@@ -48,16 +48,16 @@ MKDIR	= @MKDIR_P@
...
 ####################################################################
 # Extra (optional) source files, includes and libs
 # NOTE: EXTRA_SRC must include a solver (solver.cxx or ida_solver.cxx)
 #       and a file format (nc_format.cxx or pdb_format.cxx)
 ####################################################################
 
-EXTRA_INCS	  = @EXTRA_INCS@
-EXTRA_LIBS	  = @EXTRA_LIBS@ @OPENMP_CXXFLAGS@
+EXTRA_INCS	  = $(CPPFLAGS) @CPPFLAGS@ @EXTRA_INCS@
+EXTRA_LIBS	  = $(LIBS) @LIBS@ @EXTRA_LIBS@ @OPENMP_CXXFLAGS@
```
Comment (ZedThree) [inline make.config.in]: Not 100% this is the right thing to do, but a) we do it for `CXXFLAGS` above, and b) it seems to work

**Comment by ZedThree** [comment]
I've removed MUMPS from Travis until we get it sorted.

**Comment by ZedThree** [comment]
This is ready to go in, but needs a brave soul to give it the once over :)

The build system is still painful. We currently can't do out-of-source builds due to PVODE, so that needs to be done better. We do still set things via `CXXFLAGS` which we shouldn't do, but I haven't fixed that here. We could use pkg-config to try and get flags for various libraries, but that's not as universal as it could be. It would be good to move away from recursive makefiles, possibly even just using automake and the whole autotools bundle.

---

The parallel HDF5/NetCDF implementations are hard to test, as they need a parallel filesystem, which Travis obviously doesn't have.

The DataFile stuff needs a complete rewrite anyway, so I'm fine with dropping support for them when that happens.

File: configure.ac [hunk 2]
```diff
@@ -304,36 +307,36 @@ CXXFLAGS+=" -DBOUT_VERSION_STRING=\"\\\"$BOUT_VERSION\\\"\" -DBOUT_VERSION_DOUBL
...
 AS_IF([test "x$enable_backtrace" == "xyes" || test "x$enable_backtrace" == "xmaybe"], [
-    works=yes
-    AC_CHECK_PROG(works, addr2line, yes, no)
+    AC_CHECK_PROG([works], [addr2line], [yes], [no])
     AS_IF([test $works == no], [
-        AC_MSG_WARN("util addr2line not found")
+      AC_MSG_WARN("util addr2line not found")
     ])
-    AC_CHECK_FUNC(popen, , [works=no])
+    AC_CHECK_FUNC(popen, [works=yes], [works=no])
```
Comment (dschwoerer) [inline configure.ac]: this is a bug!
If addr2line is not working - we do not want to override the old result.

File: configure.ac [hunk 3]
```diff
@@ -304,36 +307,36 @@ CXXFLAGS+=" -DBOUT_VERSION_STRING=\"\\\"$BOUT_VERSION\\\"\" -DBOUT_VERSION_DOUBL
...
     AC_MSG_CHECKING([for execinfo.h])
-    AC_TRY_COMPILE([
-                    #include <execinfo.h>
-                   ], , , [works=no])
+    AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[
+      #include <execinfo.h>
+    ]], [[]])], [works=yes], [works=no])
```
Comment (dschwoerer) [inline configure.ac]: same as above

File: configure.ac [hunk 4]
```diff
@@ -304,36 +307,36 @@ CXXFLAGS+=" -DBOUT_VERSION_STRING=\"\\\"$BOUT_VERSION\\\"\" -DBOUT_VERSION_DOUBL
...
     AC_MSG_RESULT([$works])
     AC_MSG_CHECKING([for backtrace])
-    AC_TRY_LINK([extern "C"
-                 char backtrace();], [return backtrace();], ,
-                 [works=no])
+    AC_LINK_IFELSE([AC_LANG_PROGRAM([[
+      extern "C"
+      char backtrace();
+    ]], [[return backtrace();]])], [works=yes], [works=no])
```
Comment (dschwoerer) [inline configure.ac]: and again ...

File: configure.ac [hunk 5]
```diff
@@ -304,36 +307,36 @@ CXXFLAGS+=" -DBOUT_VERSION_STRING=\"\\\"$BOUT_VERSION\\\"\" -DBOUT_VERSION_DOUBL
...
     AC_MSG_RESULT([$works])
     AC_MSG_CHECKING([for dlfcn.h])
-    AC_TRY_COMPILE([
-                    #include <dlfcn.h>
-                   ], , , [works=no])
+    AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[
+      #include <dlfcn.h>
+    ]], [[]])], [works=yes], [works=no])
```
Comment (dschwoerer) [inline configure.ac]: one last time

File: configure.ac (hunk 2 above)
Comment (ZedThree) [reply configure.ac]: Thanks, got distracted and pushed before I was done! Now bails out early if it doesn't find a component, which is what I started doing here

Comment (dschwoerer) [reply configure.ac]: Is there a reason to set works again to yes?
It should work now - I am just curious whether this a better style?

Comment (ZedThree) [reply configure.ac]: No reason, I was just trying a few different ways of failing early and forgot to put it back to the original method. `AC_CHECK_FUNCS` (note the `S`) actually defines `HAVE_<func>` and I was initially emulating that, until I settled on the current method.

**Pull request merged**